apollo-federation/src/sources/connect/json_selection/apply_to.rs

```diff
@@ -242,53 +240,79 @@ impl ApplyTo for NamedSelection {
-                        let var_data = json!({ name: name });
-                        let mut var_path = vec![json!(name)];
-                        tail.apply_to_path(&var_data, vars, &mut var_path, errors)
+        match &self.path {
+            // If this is a KeyPath, instead of using data as given, we need to
+            // evaluate the path starting from the current value of $. To
+            // evaluate the KeyPath against data, prefix it with @. This logic
+            // supports method chaining like obj->has('a')->and(obj->has('b')),
+            // where both obj references are interpreted as $.obj.
+            PathList::Key(key, tail) => {
+                if let Some((dollar_data, dollar_path)) = vars.get("$") {
+                    let input_path_with_key = dollar_path.append(key.to_json());
+                    if let Some(child) = dollar_data.get(key.as_string()) {
+                        tail.apply_to_path(child, vars, &input_path_with_key, errors)
                     } else {
                         errors.insert(ApplyToError::new(
-                            format!("Invalid {} usage", TYPENAMES).as_str(),
-                            &[json!(var_name), json!(tail)],
+                            format!(
+                                "Property {} not found in {}",
+                                key.dotted(),
+                                json_type_name(dollar_data),
+                            )
+                            .as_str(),
+                            input_path_with_key.to_vec(),
                         ));
                         None
                     }
-                } else if let Some(var_data) = vars.get(var_name) {
-                    let mut var_path = vec![json!(var_name)];
-                    tail.apply_to_path(var_data, vars, &mut var_path, errors)
+                } else {
+                    // If $ is undefined for some reason, fall back to using data.
+                    self.path.apply_to_path(data, vars, input_path, errors)
+                }
+            }
+            path => path.apply_to_path(data, vars, input_path, errors),
+        }
+    }
+}
+
+impl ApplyTo for PathList {
+    fn apply_to_path(
+        &self,
+        data: &JSON,
+        vars: &VarsWithPathsMap,
+        input_path: &InputPath<JSON>,
+        errors: &mut IndexSet<ApplyToError>,
+    ) -> Option<JSON> {
+        match self {
+            Self::Var(var_name, tail) => {
+                if let Some((var_data, var_path)) = vars.get(var_name) {
+                    // Variables are associated with a path, which is always
+                    // just the variable name for named $variables other than $.
+                    // For the special variable $, the path represents the
+                    // sequence of keys from the root input data to the $ data.
+                    tail.apply_to_path(var_data, vars, var_path, errors)
+                } else if var_name == "@" {
```

Comment:
`@` is never in `vars`, right? If that's true, might flip those branches as a slight optimization (comparing string is faster than hashing string and looking up in map)

Reply:
https://github.com/apollographql/router/pull/5762/commits/50b8e298a3c525a45ababdbb75ca0cfc8707b37d